[PRES-366] Add density logic to player spawning

WVerlaek opened the pull request:
Adds flags that let you control player spawning, such that all players spawn in the same area and can see each other at all times.

To enable this, add worker flag `override_spawning` with value `true`, and `total_players` equal to the total number of players. All players will be spawned in a dense grid at the center of the map.

This has been adapted from https://github.com/spatialos/UnrealGDKThirdPersonShooter/pull/88, such that the density controls can be specified for the map and gamemode used in the benchmarks.


Also adds simulated player character, required for latest version of simulated players.

Zomtorg commented on Game/Source/ThirdPersonShooter/BigGym/BigGymGameMode.cpp:
```diff
@@ -17,8 +20,142 @@ ABigGymGameMode::ABigGymGameMode()
+	int NumRows = FMath::CeilToInt(FGenericPlatformMath::Sqrt(SpawnPointsNum));
+	int NumCols = FMath::CeilToInt(SpawnPointsNum / (float)NumRows);
+
+	int DistBetweenSpawnPoints = 300;
```
This seems to be in SpatialOS units, whereas in Unreal 100 units = 1 SpatialOS unit?

Zomtorg replied:
Discussed offline - it's intentional and it's so that players spawn in a grid close to each other.

m-samiec replied:
Worth clarifying if this is supposed to be meters (spatial) or centimetres (unreal)

m-samiec replied:
Worth clarifying if this is supposed to be meters (spatial) or centimetres (unreal)

WVerlaek replied:
Done

Zomtorg commented on Game/Source/ThirdPersonShooter/BigGym/BigGymGameMode.cpp:
```diff
@@ -17,8 +20,142 @@ ABigGymGameMode::ABigGymGameMode()
+	}
+
+	int NumRows = FMath::CeilToInt(FGenericPlatformMath::Sqrt(SpawnPointsNum));
+	int NumCols = FMath::CeilToInt(SpawnPointsNum / (float)NumRows);
```
Isn't `NumCols` equal to `NumRows`?

Zomtorg replied:
Answer: no, but it works either way 👍

Zomtorg commented on Game/Source/ThirdPersonShooter/BigGym/BigGymGameMode.cpp:
```diff
@@ -17,8 +20,142 @@ ABigGymGameMode::ABigGymGameMode()
+	for (int i = 0; i < SpawnPointsNum; i++)
+	{
+		int Row = i % NumRows;
+		int Col = i / NumRows;
```
Fancy 👀

m-samiec commented on Game/Source/ThirdPersonShooter/AIGym/ScaleTestableCharacter.cpp:
```diff
@@ -133,7 +135,7 @@ void AScaleTestableCharacter::BeginPlay()
 	Super::BeginPlay();
 
-	if (FParse::Param(FCommandLine::Get(), TEXT("simulatedplayer")))
+	if (FParse::Param(FCommandLine::Get(), TEXT("simulatedplayer")) || USimPlayerBPFunctionLibrary::IsSimulatedPlayer(this))
```
nit: why doesn't the static function do the args checking?

WVerlaek replied:
good point - it does do the command line arg check as well. Will remove the first clause

m-samiec commented on Game/Source/ThirdPersonShooter/BigGym/BigGymGameMode.cpp:
```diff
@@ -17,8 +20,142 @@ ABigGymGameMode::ABigGymGameMode()
+
+void ABigGymGameMode::InitGameState()
+{
+	if (ShouldUseCustomSpawning())
```
deliberately not calling super here?
If so add a comment why

WVerlaek replied:
Nope, will add the super call

m-samiec commented on Game/Source/ThirdPersonShooter/BigGym/BigGymGameMode.cpp:
```diff
@@ -17,8 +20,142 @@ ABigGymGameMode::ABigGymGameMode()
+{
+	if (FParse::Param(FCommandLine::Get(), TEXT("OverrideSpawning")))
+	{
+		if (FParse::Param(FCommandLine::Get(), TEXT("TotalPlayers=")))
```
If this doesn't exist, it shouldn't set TotalPlayers, so you should be fine to just call the get immediately.

WVerlaek replied:
Nice, changed it

m-samiec commented on Game/Source/ThirdPersonShooter/BigGym/BigGymGameMode.cpp:
```diff
@@ -17,8 +20,142 @@ ABigGymGameMode::ABigGymGameMode()
+{
+	if (FParse::Param(FCommandLine::Get(), TEXT("OverrideSpawning")))
+	{
+		if (FParse::Param(FCommandLine::Get(), TEXT("TotalPlayers=")))
```
Worth adding a log that your overriding the worker_flag setting here.

WVerlaek replied:
👍

m-samiec commented on Game/Source/ThirdPersonShooter/BigGym/BigGymGameMode.cpp:
```diff
@@ -17,8 +20,142 @@ ABigGymGameMode::ABigGymGameMode()
+void ABigGymGameMode::GenerateSpawnPoints(int SpawnPointsNum)
+{
+	// Spawn in the air above terrain obstacles.
+	int Z = 305;
```
Please make all these variables const

m-samiec commented on Game/Source/ThirdPersonShooter/BigGym/BigGymGameMode.cpp:
```diff
@@ -17,8 +20,142 @@ ABigGymGameMode::ABigGymGameMode()
+
+	if (SpawnPointsNum < 1)
+	{
+		UE_LOG(LogTemp, Error, TEXT("Can't spawn requested %d spawn points - the spawn points number should be at least 1"), SpawnPointsNum);
```
nit: please don't use LogTemp. Either define a new log category, or reuse a related log cat

WVerlaek replied:
Will create a new log category, LogTemp was used as all other BigGym classes use LogTemp as well

m-samiec commented on Game/Source/ThirdPersonShooter/BigGym/BigGymGameMode.cpp:
```diff
@@ -17,8 +20,142 @@ ABigGymGameMode::ABigGymGameMode()
+		int X = MinX + Col * DistBetweenSpawnPoints;
+		int Y = MinY + Row * DistBetweenSpawnPoints;
+
+		FActorSpawnParameters SpawnInfo;
```
```suggestion
		FActorSpawnParameters SpawnInfo{};
```

m-samiec commented on Game/Source/ThirdPersonShooter/BigGym/BigGymGameMode.cpp:
```diff
@@ -17,8 +20,142 @@ ABigGymGameMode::ABigGymGameMode()
+{
+	TArray<AActor*> ExistingSpawnPoints;
+
+	UGameplayStatics::GetAllActorsOfClass(GetWorld(), APlayerStart::StaticClass(), ExistingSpawnPoints);
```
Can't we use the SpawnPoints array here?

WVerlaek replied:
Good suggestion, that should work. Changed it to use SpawnPoints, and added a `Reset` in `GenerateSpawnPoints` to clear out any spawnpoints collected here before adding new ones

m-samiec commented on Game/Source/ThirdPersonShooter/BigGym/BigGymGameMode.cpp:
```diff
@@ -17,8 +20,142 @@ ABigGymGameMode::ABigGymGameMode()
+	// Use custom spawning with density controls
+	int32 PlayerUniqueID = Player->GetUniqueID();
+
+	UWorld* World = GetWorld();
```
This is no longer used.

WVerlaek replied:
Thanks, removed it

WVerlaek commented:
Updated this PR with an option to specify a number of player clusters with a fixed density, instead of just one cluster. Players only see other players in the same cluster.

Specify through `player_density` worker flag (which defaults to `total_players`, i.e. one cluster of players).

WVerlaek commented:
Added the option to spawn NPCs in the player clusters. Specified through `total_npcs`, it will evenly spread out `total_npcs` over all player clusters. NPCs are characters that move around in small squares.

Also fixed an issue where worker flags for the density controls were not read, as they are not available yet in `InitGameState`. It now lazily reads the worker flags when the first player is being spawned.
